lib/recursive-methods.rb

```diff
@@ -1,49 +1,74 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError.new("Number must be at least 0!") if n < 0
+  return 1 if n == 0
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2), creates a new string on each recursive call... but does s[0] add space?
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  return reverse(s[1..-1]) + s[0]
+end
+
+# Time complexity: O(n)
+# Space complexity: O(n)
+def reverse_helper(s, first, last)
+  if first < last
+    temp = s[first]
+    s[first] = s[last]
+    s[last] = temp
+    return reverse_helper(s, first + 1, last - 1)
+  end
+
+  return s
 end
 
-# Time complexity: ?
-# Space complexity: ?
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+  reverse_helper(s, 0, s.length - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return n if n == 0
+  return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2), creates a new string on each recursive call
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.length == 0
+  return false if s[0] != "(" || s[-1] != ")"
+  return nested(s[1...-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2), creates a new string on each recursive call
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+  return false if array.empty?
+  return true if array[0] == value
+  return search(array[1..-1], value)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2), creates a new string on each recursive call
 def is_palindrome(s)
-    raise NotImplementedError, "Method not implemented"
+  return true if s.length <= 1
+  return false if s[0] != s[-1]
+  return is_palindrome(s[1...-1])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log10(n))
+# Space complexity: O(log10(n))
+# Would both complexities O(log10(n*m)) since both n and m are being divided by 10 on each recursive call?
 def digit_match(n, m)
```

Comment:
This works, and the time/space complexity is O(log10 n) where n is the smaller of the two numbers.  This is because once you get the smaller number to 0, the recursion stops.